storage-proofs-update/src/vanilla.rs

```diff
@@ -252,6 +254,88 @@ where
     }
 }
 
+// Computes all `2^h` rho values for the given `phi` in the given range. Each rho corresponds to
+// a `high` value, where `high` is the `h` high bits of a node-index.
+#[derive(Debug)]
+pub struct Rhos {
+    rhos: HashMap<usize, Fr>,
+    /// The amount of right shift that is needed to get the `h` high bits of a node.
+    bits_shr: usize,
+}
+
+impl Rhos {
+    /// Generate the `rho`s for a certain number of nodes.
+    ///
+    /// Those are used for encoding.
+    pub fn new(phi: &TreeRDomain, nodes_count: usize) -> Self {
+        Self::new_range(phi, nodes_count, 0, nodes_count)
+    }
+
+    /// Generate the inverted `rho`s for a certain number of nodes.
+    ///
+    /// Those are used for decoding.
+    pub fn new_inv(phi: &TreeRDomain, nodes_count: usize) -> Self {
+        Self::new_inv_range(phi, nodes_count, 0, nodes_count)
+    }
+
+    /// Generate the `rho`s for a certain number of nodes and range.
+    ///
+    /// Those are used for encoding.
+    ///
+    /// All inputs are in number of nodes.
+    pub fn new_range(phi: &TreeRDomain, nodes_count: usize, offset: usize, num: usize) -> Self {
```

Comment:
The parameters names `nodes_count`, `offset` and `num` in these two methods are a little unclear imo.